rollup/da_syncer/syncing_pipeline.go

```diff
@@ -0,0 +1,210 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to get scroll chain abi: %w", err)
+	}
+
+	l1Client, err := rollup_sync_service.NewL1Client(ctx, ethClient, genesisConfig.Scroll.L1Config.L1ChainId, genesisConfig.Scroll.L1Config.ScrollChainAddress, scrollChainABI)
+	if err != nil {
+		return nil, err
+	}
+
+	blobClientList := blob_client.NewBlobClientList()
+	if config.BlobScanAPIEndpoint != "" {
+		blobClientList.AddBlobClient(blob_client.NewBlobScanClient(config.BlobScanAPIEndpoint))
+	}
+	if config.BlockNativeAPIEndpoint != "" {
+		blobClientList.AddBlobClient(blob_client.NewBlockNativeClient(config.BlockNativeAPIEndpoint))
+	}
+	if blobClientList.Size() == 0 {
+		log.Crit("DA syncing is enabled but no blob client is configured. Please provide at least one blob client via command line flag.")
+	}
+
+	dataSourceFactory := NewDataSourceFactory(blockchain, genesisConfig, config, l1Client, blobClientList, db)
+	syncedL1Height := l1DeploymentBlock - 1
+	from := rawdb.ReadDASyncedL1BlockNumber(db)
+	if from != nil {
+		syncedL1Height = *from
+	}
+
+	daQueue := NewDAQueue(syncedL1Height, dataSourceFactory)
+	batchQueue := NewBatchQueue(daQueue, db)
+	blockQueue := NewBlockQueue(batchQueue)
+	daSyncer := NewDASyncer(blockchain)
+
+	ctx, cancel := context.WithCancel(ctx)
+	return &SyncingPipeline{
+		ctx:               ctx,
+		cancel:            cancel,
+		expBackoff:        backoff.NewExponential(100*time.Millisecond, 10*time.Second, 100*time.Millisecond),
+		wg:                sync.WaitGroup{},
+		l1DeploymentBlock: l1DeploymentBlock,
+		db:                db,
+		blockchain:        blockchain,
+		blockQueue:        blockQueue,
+		daSyncer:          daSyncer,
+	}, nil
+}
+
+func (s *SyncingPipeline) Step() error {
+	block, err := s.blockQueue.NextBlock(s.ctx)
+	if err != nil {
+		return err
+	}
+	err = s.daSyncer.SyncOneBlock(block)
+	return err
+}
+
+func (s *SyncingPipeline) Start() {
+	log.Info("sync from DA: starting pipeline")
+
+	s.wg.Add(1)
+	go func() {
+		s.mainLoop()
+		s.wg.Done()
+	}()
+}
+
+func (s *SyncingPipeline) mainLoop() {
+	stepCh := make(chan struct{}, 1)
+	var delayedStepCh <-chan time.Time
+	var resetCounter int
+
+	// reqStep is a helper function to request a step to be executed.
+	// If delay is true, it will request a delayed step with exponential backoff, otherwise it will request an immediate step.
+	reqStep := func(delay bool) {
+		if delay {
+			if delayedStepCh == nil {
+				delayDur := s.expBackoff.NextDuration()
+				delayedStepCh = time.After(delayDur)
+				log.Debug("requesting delayed step", "delay", delayDur, "attempt", s.expBackoff.Attempt())
+			} else {
+				log.Debug("ignoring step request because of ongoing delayed step", "attempt", s.expBackoff.Attempt())
+			}
+		} else {
+			select {
+			case stepCh <- struct{}{}:
+			default:
+			}
+		}
+	}
+
+	// start pipeline
+	reqStep(false)
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case <-s.ctx.Done():
+			return
```

Comment:
can we merge these 2 "select"s?

Reply:
We might be able to merge this now if each individual case handles a shutdown (done context) correctly. Before this wasn't the case.